### PR TITLE
Add Null Type in TS and Kotlin. Add Null Constant and Undefined Type in Typescript

### LINKED ISSUE #4: Null & Undefined Types

### PR DESCRIPTION
**Pull request by alexisthedev**

### CODE REVIEW

File: src/ir/groovy_types.py [hunk 1]
```diff
@@ -72,6 +72,10 @@ def get_primitive_types(self):
             BooleanType(primitive=True)
         ]
 
+    def get_null_type(self):
+        # FIXME
+        raise Exception("Groovy does not support null types")
```
Comment (theosotr) [inline src/ir/groovy_types.py:76]: remove FIXME

File: src/ir/java_types.py [hunk 2]
```diff
@@ -62,6 +62,10 @@ def get_big_integer_type(self):
     def get_function_type(self, nr_parameters=0):
         return FunctionType(nr_parameters)
 
+    def get_null_type(self):
+        # FIXME
```
Comment (theosotr) [inline src/ir/java_types.py]: same here

File: src/ir/kotlin_types.py [hunk 3]
```diff
@@ -123,6 +126,17 @@ def get_builtin_type(self):
         return bt.Number
 
 
+class NullType(KotlinBuiltin):
```
Comment (theosotr) [inline src/ir/kotlin_types.py]: I think the parent of `NullType` should be `AnyType`.

File: src/ir/typescript_types.py [hunk 4]
```diff
@@ -181,6 +192,28 @@ def get_name(self):
         return super().get_name()
 
 
+class NullType(TypeScriptBuiltin):
+    def __init__(self, name="null", primitive=False):
+        super().__init__(name, primitive)
```
Comment (theosotr) [inline src/ir/typescript_types.py]: I think the object type should be the supertype of `NullType`.

File: src/ir/typescript_types.py [hunk 5]
```diff
@@ -181,6 +192,28 @@ def get_name(self):
...
+    def box_type(self):
+        return NullType(self.name)
+
+    def get_name(self):
+        return 'null'
+
+
+class UndefinedType(TypeScriptBuiltin):
+    def __init__(self, name="undefined", primitive=False):
+        super().__init__(name, primitive)
```
Comment (theosotr) [inline src/ir/typescript_types.py]: I think the object type should be the supertype of `UndefinedType`.

**Review by theosotr — CHANGES_REQUESTED** [review]

**Comment by theosotr** [comment]
Just to be sure. What's the supertype of all types in TypeScript? I think it's not `any` because this type disables type checking. So, what's the supertype of e.g., `number`?

**Comment by alexisthedev** [comment]
> Just to be sure. What's the supertype of all types in TypeScript? I think it's not `any` because this type disables type checking. So, what's the supertype of e.g., `number`?

The supertype of all types in TS is the `Any` type.

From what I gather from the docs, it holds that role alongside `unknown`, [with the only difference between the two being that `unknown` can **only** be assigned to any.](https://www.typescriptlang.org/docs/handbook/type-compatibility.html#any-unknown-object-void-undefined-null-and-never-assignability)

This is not a surprise, as this is the default behavior of JavaScript (everything is of type any).

**Comment by theosotr** [comment]
> > Just to be sure. What's the supertype of all types in TypeScript? I think it's not `any` because this type disables type checking. So, what's the supertype of e.g., `number`?
> 
> The supertype of all types in TS is the `Any` type.
> 
> From what I gather from the docs, it holds that role alongside `unknown`, [with the only difference between the two being that `unknown` can **only** be assigned to any.](https://www.typescriptlang.org/docs/handbook/type-compatibility.html#any-unknown-object-void-undefined-null-and-never-assignability)
> 
> This is not a surprise, as this is the default behavior of JavaScript (everything is of type any).

So, If the supertype in TS is `any`, is there any reason why we chose `ObjectType` as the supertype of `NumberType`, `NullType`, etc?

**Comment by alexisthedev** [comment]
> So, If the supertype in TS is `any`, is there any reason why we chose `ObjectType` as the supertype of `NumberType`, `NullType`, etc?

Since 'any' in typescript bypasses typechecking having a 'AnyType' class in typescript_types.py would have no use as we would never pass AnyType to the generator. Hence we moved on with just using ObjectType.

File: src/ir/typescript_types.py [hunk 6]
```diff
@@ -48,29 +48,42 @@ def get_primitive_types(self):
...
             BooleanType(primitive=True),
-            BigIntegerType(primitive=True)
+            BigIntegerType(primitive=True),
+            NullType(primitive=True),
+            UndefinedType(primitive=True)
         ]
 
     def get_integer_type(self):
-        return NumberType(primitive=False)
+        return NumberType(primitive=True)
```
Comment (theosotr) [inline src/ir/typescript_types.py]: Why `primitive=True`?

File: src/ir/typescript_types.py [hunk 7]
```diff
@@ -48,29 +48,42 @@ def get_primitive_types(self):
...
 
     def get_byte_type(self):
-        return NumberType(primitive=False)
+        return NumberType(primitive=True)
```
Comment (theosotr) [inline src/ir/typescript_types.py]: Same here

**Pull request merged**